woocommerce/woocommerce-gutenberg-products-block
# Break down Store API controllers into specific routes

#### What is this? 

This PR converts our Store API controllers into classes for individual routes.

#### Why do this?

#1938 needs another route under `/cart`. The `/cart/order` route requires many methods so adding this to the existing cart controller would make things difficult to maintain.

Each controller was responsible for all endpoints and sub-routes under a route. For example, `/cart` had a single class handling the index route, apply coupon, update item, update shipping and so on. This lead to a monster of a class doing many things at once. 

Having single controllers also opened up the possibility of accidentally implementing incorrect schema for a route which almost happened recently when adding routes under `/cart`.

The code in this PR effectively:

- Abstracts out routes to create shared methods
- Removes the reliance on the core controller class
- Enforces a single route per class. A route can however be made up of multiple endpoints (PUT POST GET etc)
- Enforces a single **schema** per route
- Makes each class smaller and more focussed 
- Makes it simpler to add new routes or endpoints under an existing route

Whilst refactoring this, I removed the now unused `/customer` and `/shipping-rates` routes.

A follow up PR is incoming for the changes in #1938.

### How to test the changes in this Pull Request:

1. Smoke test various Store API endpoints
2. Test the Product Block and cart since those use Store API endpoints
3. Ensure tests continue to pass
